### PR TITLE
chore: storage method to sign tx with custom override

### PR DESCRIPTION
**Pull request by r4mmer**
### Acceptance Criteria

- Allow the user to override the tx signature method.
- Default to `transactionUtils.signTransaction`, our main signature method.

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/new/wallet.js
Comment (r4mmer) [inline src/new/wallet.js:1574]: note: These are just to remove unnecessary awaits.

File: src/new/wallet.js
Comment (r4mmer) [inline src/new/wallet.js:2363]: note: This will allow us to sign transactions for P2SH wallets using either the bitcore signing from the storage mainKey or the custom signing method.

File: src/storage/storage.ts
Comment (r4mmer) [inline src/storage/storage.ts:97]: note: This is the core feature of this review.

Any signing on the wallet should be done from this method, this will allow us to maintain support for external signing methods going forward.

File: src/storage/storage.ts
Comment (r4mmer) [inline src/storage/storage.ts:166]: note: We only got the pubkey from the derived mainKey, this does not work for external signing where we only have the xpubkey.

File: src/utils/transaction.ts
Comment (r4mmer) [inline src/utils/transaction.ts:126]: note: This new util is the default signing method for the wallet-lib for now.

File: src/storage/storage.ts
Comment (tuliomir) [reply src/storage/storage.ts:97]: question: Why does codecov warns us of lack of coverage here, since this is now the main form of signing transactions on the lib? Shouldn't this be covered by most of all of our transaction tests?

**Review by tuliomir — COMMENTED** [review]
thought(blocking): It troubles me that we're changing a central protection ( read only guards ) and removing the tests that would assure us it's working as intended later.

I'd suggest splitting this feature differently:
- First with a refactor for the transaction utils with all new signature methods. All tests would have to be green still
- Then adding a flag to allow for a full wallet with external cryptography ( the scope of #619 , as I understand ). The tests could be adapted accordingly, with almost no new tests needed.
- Lastly, implement the feature to actually initialize the wallet without any `xPriv` and having an external signing method. New tests would be made for this new feature.

What do you think of this approach?

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:97]: The integration test does not upload the coverate artifact when it fails.

The tests are actually passing, its just the wallet facade coverate that dropped slightly but it will be back once the xpub guard PR is merged

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:165]: This is meant to be used to get the pubkey of the path `m/44'/280'/0'/0/index` where index is the address index.

When we sign the transaction with an external method we still need the actual pubkey to build the input data (on both multisig and p2pkh) but we cannot use the main key to derive it (which is the method we use on the lib).

Also, If the public key is generated during the initial address derivation we will use it (on the `addressInfo.publicKey`).

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:154]: Done

File: src/new/wallet.js
Comment (r4mmer) [reply src/new/wallet.js]: Renamed to `sigInfoArray`

File: src/storage/storage.ts
Comment (tuliomir) [reply src/storage/storage.ts:181]: Is there any downside to reusing the `deriveAddressFromXPubP2PKH()` function here instead of deriving manually?

https://github.com/HathorNetwork/hathor-wallet-lib/blob/91118d5335c2b63afb559e5deddc7a81f73cbc13/src/utils/address.ts#L27-L31 ?

Comment (r4mmer) [reply src/storage/storage.ts:181]: This is meant to be used by both p2pkh and p2sh wallets.
If you really want to minimize repetition the best approach would be to create an util `getAddressPubkeyFromXpub` and use it on both the `deriveAddressFromXPubP2PKH` and this new method.
But I think this is creates a new method for a very simple derive operation.

File: src/new/wallet.js
Comment (tuliomir) [inline src/new/wallet.js:525]: note(non-blocking): Codecov is raising one case that was not tested: having invalid p2sh signatures.

Since P2SH transactions involve message exchange and direct user input I think it's something actually _plausible_ to happen. Maybe we could open an issue to remind us of this test case.

File: src/storage/storage.ts
Comment (tuliomir) [inline src/storage/storage.ts:92]: question: Would it be too much of an effort to create a `txSignFunc` and test it? The best case would be on an integration test, but I believe just a unit test would be enough for now.

That way we would ensure the override process is actually working.

File: src/storage/storage.ts
Comment (tuliomir) [inline src/storage/storage.ts:171]: chore: We have no coverage on the new `getAddressPubkey` method. I feel this is important to validate the new feature.

File: src/storage/storage.ts
Comment (tuliomir) [inline src/storage/storage.ts:101]: suggestion: This method does not sign the transaction, but only returns an array of signatures for another method to inject it into the `Transaction` object.

I'd suggest renaming this method to `getTxSignatures(tx, pinCode)` to better reflect that it is a pure function.

File: src/utils/transaction.ts
Comment (pedroferreira1) [inline src/utils/transaction.ts:148]: We used to have `xpriv.publicKey.toBuffer()` for this. Is there any difference between both codes?

File: src/storage/storage.ts
Comment (pedroferreira1) [inline src/storage/storage.ts:170]: Are you using this method somewhere?

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:92]: `getSignatureForTx` is the perfect example, it fetches/derives the mainKey and uses the storage to iterate on the inputs that are owned by the wallet, signing only those that belong to the wallet.

Any other test method would essentially need to have to do the same.
During tests what I would do is create a function that prints something to stdout and call `getSignatureForTx`, so I would know it called the correct method and that it works as expected.

File: src/types.ts
Comment (r4mmer) [reply src/types.ts:21]: Yes, the initial implementation would receive the data to sign and the address index.
During the implementation I changed since this is not ideal, because it would require retrieving the key from storage and decrypting + signing it for each input instead of fetching once and signing multiple times.

This would be bad specially for HSM since we would need to start a session for each address key, even when we have 2 inputs for the same address.

Comment (r4mmer) [reply src/types.ts:21]: I'll fix the docstring

File: src/storage/storage.ts
Comment (tuliomir) [reply src/storage/storage.ts:92]: I'm not sure if I understand: `getSignatureForTx` is an example of signing function - actually the default one used by the lib, right?

CoveCov is informing that we have not tested this line `92`. As I understand, to increase this coverage in unit tests we would just need to call `setTxSign` with a mocked function and ensure it's actually invoked whenever a signature is required within the scope of this storage/wallet.

Integration tests for an actual valid external signature could be left to another moment.

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:91]: Having the storage hold this method makes more sense since this method will be responsible for fetching the private key, identifying which inputs to sign (and the address indexes to use) then signing.
The storage is expected to hold the private keys and addresses information.
In the future the wallets can just override this method on the storage and all internals of the wallet-lib will work.

It also makes the wallet-lib implementation simpler since some internal utilities use the storage instance and not the wallet facade instance, we would need to change them and take care to always pass the wallet facade anywhere we would need to sign anything.

Comment (r4mmer) [reply src/storage/storage.ts:91]: If you think we should, we can also have a facade method to set this on the storage.
What do you think?

File: src/utils/transaction.ts
Comment (r4mmer) [reply src/utils/transaction.ts:148]: No, internally `toBuffer()` is the same method as `toDER()`.
[reference](https://github.com/bitpay/bitcore/blob/v10.0.28/packages/bitcore-lib/lib/publickey.js#L327)

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:170]: Not in the wallet-lib but HSM custom signature method will require this.

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:101]: Done

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:171]: Added tests

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts:92]: Created an integration test showcasing this.
It will also increase the coverage since it tests the custom signature method case.

File: src/new/wallet.js
Comment (r4mmer) [reply src/new/wallet.js:525]: Can you create an issue for this?
I think its important, but it is beyond the scope of this PR.

File: src/new/wallet.js
Comment (pedroferreira1) [inline src/new/wallet.js:2371]: Why did you remove this?

File: src/storage/storage.ts
Comment (pedroferreira1) [reply src/storage/storage.ts]: I agree that it's good to have it on storage because of your explanation but I think it's good also to have it on the facade. 

I think that any external client that wants to use this method shouldn't have to call the storage. So I like your suggestion, to create a method in the facade that calls this storage method.

File: src/new/wallet.js
Comment (r4mmer) [reply src/new/wallet.js:2371]: Added the pin check back, but the rest was moved to the new signing method.

File: src/storage/storage.ts
Comment (r4mmer) [reply src/storage/storage.ts]: Done

**Review by pedroferreira1 — APPROVED** [review]

File: src/new/wallet.js
Comment (tuliomir) [reply src/new/wallet.js:540]: Opened #618 requesting this test

**Pull request merged**
